File: server/src/main/com/broll/gainea/server/core/goals/impl/all/G_StackUnits.java
package com.broll.gainea.server.core.goals.impl.all;

import com.broll.gainea.misc.RandomUtils;
import com.broll.gainea.server.core.GameContainer;
import com.broll.gainea.server.core.bot.strategy.GoalStrategy;
import com.broll.gainea.server.core.goals.CustomOccupyGoal;
import com.broll.gainea.server.core.goals.GoalDifficulty;
import com.broll.gainea.server.core.map.Area;
import com.broll.gainea.server.core.objects.Unit;
import com.broll.gainea.server.core.player.Player;
import com.google.common.collect.Sets;

public class G_StackUnits extends CustomOccupyGoal {
    private final static int COUNT = 6;
    private Area area;

    public G_StackUnits() {
        super(GoalDifficulty.EASY, null);
    }

    @Override
    public boolean init(GameContainer game, Player player) {
        this.game = game;
        this.player = player;
        area = RandomUtils.pickRandom(game.getMap().getAllAreas());
        text = "Besetze " + area.getName() + " mit mindestens " + COUNT + " Einheiten";
        setExpansionRestriction(area.getContainer().getExpansion().getType());
        setProgressionGoal(COUNT);
        return true;
    }

    @Override
    public void check() {
        int playerUnits = (int) area.getInhabitants().stream().filter(it -> it.getOwner() == player && it instanceof Unit).count();
        updateProgression(playerUnits);
        if (playerUnits >= COUNT) {
            success();
        }
    }

    @Override
    public void botStrategy(GoalStrategy strategy) {
        strategy.setRequiredUnits(COUNT);
        strategy.updateTargets(Sets.newHashSet(area));
    }

}

File: server/src/main/com/broll/gainea/server/core/goals/impl/all/G_StackUnits.kt
package com.broll.gainea.server.core.goals.impl.all

import com.broll.gainea.server.core.bot.strategy.GoalStrategy
import com.broll.gainea.server.core.goals.GoalDifficulty
import com.broll.gainea.server.core.goals.OccupyGoal
import com.broll.gainea.server.core.map.Area
import com.broll.gainea.server.core.objects.Unit

class G_StackUnits : OccupyGoal(GoalDifficulty.EASY, "") {
    private lateinit var area: Area

    init {
        autoCheckProgressions = false
        progressionGoal = COUNT
        libraryText = text("X")
    }

    private fun text(area: String) =
        "Besetze $area mit mindestens $COUNT Einheiten"

    override fun initOccupations() {
        area = game.map.allAreas.random()
        text = text(area.name)
        condition(occupy(area), {
            it.inhabitants.count { it.owner == player && it is Unit } >= COUNT
        })
    }

    override fun check() {
        val playerUnits = area.inhabitants.count { it.owner === player && it is Unit }
        updateProgression(playerUnits)
        super.check()
    }

    override fun botStrategy(strategy: GoalStrategy) {
        strategy.setRequiredUnits(COUNT)
        strategy.updateTargets(setOf(area))
    }


    companion object {
        private const val COUNT = 6
    }
}
